src/open_inwoner/utils/tests/test_migrations.py

```diff
@@ -37,18 +35,71 @@ def setUp(self):
         # Reverse to the original migration
         executor.migrate(self.migrate_from)
 
-        self.setUpBeforeMigration(old_apps)
-
+    def apply_migration_to(self):
         # Run the migration to test
         executor = MigrationExecutor(connection)
         executor.loader.build_graph()  # reload.
         executor.migrate(self.migrate_to)
+        self.apps = executor.loader.project_state(self.migrate_to).apps
 
+    def restore_constraint_checks(self):
         # Restore constraint checks
         with connection.cursor() as cursor:
             cursor.execute("SET CONSTRAINTS ALL DEFERRED")
 
-        self.apps = executor.loader.project_state(self.migrate_to).apps
-
     def setUpBeforeMigration(self, apps):
         pass
+
+
+class TestSuccessfulMigrations(TestMigrationsBase):
+    """Test a successful migration.
+
+    Set the class attributes `app`, `migration_from`, and `migrate_to`. You
+    can specify your pre-migration state in `setUpBeforeMigration()`, and
+    consequently assert against the resulting state (in your test, do
+    not import models directly but use:
+
+    ```
+    MyModel = self.apps.get_model(self.app, "MyModel")
+    ```
+    """
+
+    def setUp(self):
+        self.revert_to_migrate_from()
+        self.setUpBeforeMigration(self.old_apps)
+        self.apply_migration_to()
+        self.restore_constraint_checks()
+
+
+class TestFailingMigrations(TestMigrationsBase):
+    """Test a migration which is expected to fail.
+
+    Set the class attributes `app`, `migration_from`, and `migrate_to`. You
+    can specify your pre-migration state in `setUpBeforeMigration()`, though
+    you should not import models directly, but rather use:
+
+    ```
+    MyModel = self.apps.get_model(self.app, "MyModel")
+    ```
+
+    In your test, you can attempt to migration using `self.attempt_migration()`
+    and assert against the expected exception:
+    ```
+    def test_migration_should_fail(self):
+        # Setup failure conditions
+        self.assertRaises(SomeError):
+            self.attempt_migration()
+    ```
+    """
+
+    def setUp(self):
+        self.revert_to_migrate_from()
+        self.setUpBeforeMigration(self.old_apps)
```

Comment:
`setUpBeforeMigration` has `cursor.execute("SET CONSTRAINTS ALL IMMEDIATE")`, which is executed independently of whether the migration succeeds. Do we not need to restore the constraints?